### PR TITLE
Feature/TR-701/Convert code to ES6

### PR DESCRIPTION
**Pull request by jsconan**
Related to: https://oat-sa.atlassian.net/browse/TR-701

Preliminary work on the extension to modernise the code and convert it to ECMAScript 2015.

There is a lot of files there, sorry. But the changes are only on the code flavour. There should not be any change in the behaviour nor in the visual of the diagnostic tool.

**What to check/test?**
- look at the code change, it should be full ES6 now. But unit tests did not change though.
- unit tests must all pass
    ```
    cd tao/views/build
    npx grunt connect:dev taoclientdiagnostictest
    ```
- the application must work as previously

**How to run the application?**
First of all, the extension needs to be installed, obviously:
```
php tao/scripts/installExtension.php taoClientDiagnostic
```

There are actually 2 modes:
- anonymous diagnostic
- logged on diagnostic

To access the diagnostic from the anonymous mode, open this url in your browser:
`http://YOUR_HOST/taoClientDiagnostic/CompatibilityChecker/index`
image

Then click the `Test system compatibility` button.

To access the diagnostic from the logged on mode, first you need to log in, as a user having the role `Readiness checker`.
image

Then open this url in your browser:
`http://YOUR_HOST/taoClientDiagnostic/Diagnostic/index`

A data grid will show first, then you can click the `Launch readiness check` button.
image

The diagnostic part is similar to the anonymous part, with the difference there is a field for the workstation's name.
image

**Note:** By default, not all options nor all testers are enabled. They can be activated from the config file at `config/taoClientDiagnostic/clientDiag.conf.php`.

The testers are all listed in this config and some may be disabled. To enable them all, you need to set the `enabled` property to `true` for all of them.

The following options could also be turned on for a more comprehensive test:
- `requireSchoolName`
- `requireSchoolId`
- `validateSchoolName`

It will add some fields to the diagnostic screen:
image

### CODE REVIEW

File: views/js/tools/getStatus.js [hunk 1]
```diff
@@ -13,50 +13,46 @@
...
-            if (!_.isArray(thresholds)) {
+            if (!Array.isArray(thresholds)) {
                 thresholds = [thresholds];
             }
 
-            len = thresholds.length;
-            for (i = 0; i < len; i++) {
-                step = thresholds[i];
+            let feedback;
+            const len = thresholds.length >>> 0;
```
Comment (marpesia) [inline views/js/tools/getStatus.js:53]: Is this ok? It is first time I see `>>>`

**Review by marpesia — CHANGES_REQUESTED** [review]

File: views/js/tools/getStatus.js (hunk 1 above)
Comment (jsconan) [reply views/js/tools/getStatus.js:53]: Local optimisation often used for loop over arrays, forcing the number to become an `Int32`.
You can see an example here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach#polyfill

File: views/js/tools/diagnostic/diagnostic.js [hunk 2]
```diff
@@ -310,254 +305,247 @@ define([
...
-                var self = this;
-
-                /**
-                 * Default launcher
-                 */
-                var launch = function launch() {
-                    runDiagnostics();
-                };
+            .on('render', function onDiagnosticRender() {
+                const self = this;
```
[the remark this answers is not in the document]
Comment (jsconan) [reply views/js/tools/diagnostic/diagnostic.js:490]: good catch

**Review by marpesia — APPROVED** [review]
Good job!

**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ]  Commits messages are meaningful
- [ ]  Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by meefox — APPROVED** [review]
#
**Review Checklist**
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Pull request's target is not `master`
- [ ] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Pull request merged**